community/index/src/main/java/org/neo4j/index/internal/gbptree/TreeNode.java

```diff
@@ -30,42 +30,22 @@
 /**
  * Methods to manipulate single tree node such as set and get header fields,
  * insert and fetch keys, values and children.
- * <p>
- * DESIGN
- * <p>
- * Using Separate design the internal nodes should look like
- * <pre>
- * # = empty space
- *
- * [                                   HEADER   82B                           ]|[   KEYS   ]|[     CHILDREN      ]
- * [NODETYPE][TYPE][GENERATION][KEYCOUNT][RIGHTSIBLING][LEFTSIBLING][SUCCESSOR]|[[KEY]...##]|[[CHILD][CHILD]...##]
- *  0         1     2           6         10            34           58          82
- * </pre>
- * Calc offset for key i (starting from 0)
- * HEADER_LENGTH + i * SIZE_KEY
- * <p>
- * Calc offset for child i
- * HEADER_LENGTH + SIZE_KEY * MAX_KEY_COUNT_INTERNAL + i * SIZE_CHILD
- * <p>
- * Using Separate design the leaf nodes should look like
- *
- * <pre>
- * [                                   HEADER   82B                           ]|[    KEYS  ]|[   VALUES   ]
- * [NODETYPE][TYPE][GENERATION][KEYCOUNT][RIGHTSIBLING][LEFTSIBLING][SUCCESSOR]|[[KEY]...##]|[[VALUE]...##]
- *  0         1     2           6         10            34           58          82
- * </pre>
- *
- * Calc offset for key i (starting from 0)
- * HEADER_LENGTH + i * SIZE_KEY
- * <p>
- * Calc offset for value i
- * HEADER_LENGTH + SIZE_KEY * MAX_KEY_COUNT_LEAF + i * SIZE_VALUE
- *
- * @param <KEY> type of key
- * @param <VALUE> type of value
  */
-class TreeNode<KEY,VALUE>
+abstract class TreeNode<KEY,VALUE>
 {
+    enum Type
+    {
+        LEAF,
+        INTERNAL
+    }
+
+    enum Overflow
+    {
+        YES,
+        NO,
+        NEED_DEFRAG
```

Comment:
I was thinking... `NEED_DEFRAG` means `YES`, but with a condition that it must be defragged first, right? Could it be called `YES_AFTER_DEFRAG` instead?

Reply:
`NEED_DEFRAG` mean `NO_BUT_YOU_NEED_TO_DO_SOME_CLEANUP_FIRST`. They all answer the question "will I get an OVERFLOW if I try to insert this item?". Will make naming more clear.